src/compiler/crystal/codegen/cache_dir.cr

```diff
@@ -0,0 +1,63 @@
+module Crystal
+  # Manages cache files in the ".crystal" directory.
+  #
+  # For each compiled program a directory is created in the cache
+  # that stores .bc and .o files that could possibly be reused
+  # from a previous compilation.
+  #
+  # To keep the cache dir small, only the 10 most recently used
+  # directories are kept. We use the directory's modification
+  # time for this.
+  class CacheDir
+    def initialize(@dir : String)
+    end
+
+    # Returns the directory where cache files related to the
+    # given sources will be stored. The directory will be
+    # created if it doesn't exist.
+    def directory_for(sources : Array(Compiler::Source))
+      directory_for(sources.first.filename)
+    end
+
+    # Returns the directory where cache files related to the
+    # given filenames will be stored. The directory will be
+    # created if it doesn't exist.
+    def directory_for(filename : String)
+      name = filename.gsub('/', '-')
+      while name.starts_with?('-')
+        name = name[1..-1]
+      end
+      output_dir = File.join(@dir, name)
+      Dir.mkdir_p(output_dir)
+      output_dir
+    end
+
+    # Keeps the 10 most recently used directories in the cache,
+    # and removes all others. This also removes non-directory
+    # files inside the cache directory (temporary executables
+    # resulting from `crystal run` or `run` macro calls).
+    def cleanup
+      cleanup_dirs
+      cleanup_files
+    end
+
+    private def cleanup_dirs
+      Dir.entries(@dir)
+         .reject { |name| name == "." || name == ".." }
+         .map! { |name| File.join(@dir, name) }
+         .select { |dir| Dir.exists?(dir) }
+         .sort_by! { |dir| File.stat(dir).mtime rescue Time.epoch(0) }
```

Comment:
You are right, but in a non-optimized compiler with more than 10 directories it took about `0.003` seconds or something like that, and since the number will be capped to 10 I think keeping the code clear is more important here (some intermediate arrays are created too, but since this isn't going to be a bottleneck we can use these nice features :-)).